File: scripts/manual-tests/test_audio_batch_preview.py
#!/usr/bin/env python3
"""
End-to-end manual test for DCT-51: audio_url dataset schema field +
`aitaskbuilder batch preview` command.

Builds the CLI, creates a dataset with an audio_url field, uploads sample
data, creates a batch, adds instructions, sets up the batch, then previews it.

Usage:
    python3 scripts/manual-tests/test_audio_batch_preview.py [workspace_id]

Requires PROLIFIC_TOKEN / PROLIFIC_URL set in the environment for the target
API. Run from anywhere; paths are resolved relative to the repo root.
"""

import os
import re
import subprocess
import sys
import tempfile

REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
CLI_BINARY = os.path.join(tempfile.gettempdir(), "prolific-cli")
DEFAULT_WORKSPACE_ID = "679271425fe00981084a5f58"  # DCT Workspace


def run(args, check=True):
    print(f"\n$ {' '.join(args)}")
    result = subprocess.run(args, capture_output=True, text=True)
    print(result.stdout)
    if result.stderr:
        print(result.stderr, file=sys.stderr)
    if check and result.returncode != 0:
        print(f"Command failed with exit code {result.returncode}", file=sys.stderr)
        sys.exit(result.returncode)
    return result.stdout


def extract_field(output, label):
    match = re.search(rf"^{re.escape(label)}:\s*(\S+)", output, re.MULTILINE)
    if not match:
        print(f"Could not find '{label}:' in output:\n{output}", file=sys.stderr)
        sys.exit(1)
    return match.group(1)


def build_cli():
    print("Building CLI...")
    run(["go", "build", "-o", CLI_BINARY, "."], check=True)


def main():
    os.chdir(REPO_ROOT)
    workspace_id = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_WORKSPACE_ID

    build_cli()

    # 1. Create dataset with an audio_url field
    schema = (
        '{"fields":{"question":{"type":"text","label":"Question"},'
        '"clip":{"type":"audio_url","label":"Audio clip"}}}'
    )
    output = run(
        [
            CLI_BINARY, "aitaskbuilder", "dataset", "create",
            "-n", "Audio URL Test Dataset",
            "-w", workspace_id,
            "--strict",
            "--schema", schema,
        ]
    )
    dataset_id = extract_field(output, "ID")
    print(f"Created dataset: {dataset_id}")

    # 2. Upload sample data containing audio URLs
    csv_path = os.path.join(tempfile.gettempdir(), "audio-dataset.csv")
    with open(csv_path, "w") as f:
        f.write(
            "question,clip\n"
            '"What emotion is being expressed?","https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3"\n'
            '"Transcribe the spoken word.","https://www.w3schools.com/html/horse.mp3"\n'
        )

    dataset_id = "019f6123-1fdc-7518-a3ec-75b830c5a5f4"
    run([CLI_BINARY, "aitaskbuilder", "dataset", "upload", "-d", dataset_id, "-f", csv_path])

    # 3. Check dataset status (informational; may need polling until READY)
    run([CLI_BINARY, "aitaskbuilder", "dataset", "check", "-d", dataset_id], check=False)

    # 4. Create batch linked to the dataset
    output = run(
        [
            CLI_BINARY, "aitaskbuilder", "batch", "create",
            "-n", "Audio URL Test Batch",
            "-w", workspace_id,
            "-d", dataset_id,
            "--task-name", "Audio Review Task",
            "--task-introduction", "Listen to the audio clip and answer the question.",
            "--task-steps", "1. Listen to the clip\\n2. Answer the question",
        ]
    )
    batch_id = extract_field(output, "ID")
    print(f"Created batch: {batch_id}")

    # 5. Add instructions (required before setup)
    run(
        [
            CLI_BINARY, "aitaskbuilder", "batch", "instructions",
            "-b", batch_id,
            "-j", '[{"type":"free_text","created_by":"Cem","description":"Please describe what you heard."}]',
        ]
    )

    # 6. Setup the batch
    run(
        [
            CLI_BINARY, "aitaskbuilder", "batch", "setup",
            "-b", batch_id,
            "-d", dataset_id,
            "--tasks-per-group", "1",
        ]
    )

    # 7. Preview the batch (the new command under test)
    run([CLI_BINARY, "aitaskbuilder", "batch", "preview", "-b", batch_id])

    print("\nDone. Review output above for correctness.")


if __name__ == "__main__":
    main()

File: scripts/manual-tests/test_audio_batch_preview.go
package main

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"
)

const defaultWorkspaceID = "679271425fe00981084a5f58" // DCT Workspace

var fieldPattern = regexp.MustCompile(`(?m)^([A-Za-z ]+):\s*(\S+)$`)

func main() {
	repoRoot, err := repoRoot()
	if err != nil {
		fatal(err)
	}

	workspaceID := defaultWorkspaceID
	if len(os.Args) > 1 {
		workspaceID = os.Args[1]
	}

	cliBinary := filepath.Join(os.TempDir(), "prolific-cli")
	if err := buildCLI(repoRoot, cliBinary); err != nil {
		fatal(err)
	}

	schema := `{"fields":{"question":{"type":"text","label":"Question"},"clip":{"type":"audio_url","label":"Audio clip"}}}`
	batchItems := `[{"rows":[{"columns":[{"items":[{"type":"dataset_field","field":"question"},{"type":"dataset_field","field":"clip"},{"type":"free_text","description":"Please describe what you heard."}]}]}]}]`

	// Create a V4 dataset whose schema includes an audio_url field.
	output, err := run(repoRoot, cliBinary,
		"aitaskbuilder", "dataset", "create",
		"-n", "Audio URL Test Dataset",
		"-w", workspaceID,
		"--strict",
		"--schema", schema,
	)
	if err != nil {
		fatal(err)
	}

	datasetID, err := extractField(output, "ID")
	if err != nil {
		fatal(err)
	}
	fmt.Printf("Created dataset: %s\n", datasetID)

	csvPath := filepath.Join(os.TempDir(), "audio-dataset.csv")
	csvContents := strings.Join([]string{
		"question,clip",
		`"What emotion is being expressed?","https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3"`,
		`"Transcribe the spoken word.","https://www.w3schools.com/html/horse.mp3"`,
		"",
	}, "\n")
	if err := os.WriteFile(csvPath, []byte(csvContents), 0o600); err != nil {
		fatal(fmt.Errorf("failed to write CSV fixture: %w", err))
	}

	invalidCSVPath := filepath.Join(os.TempDir(), "audio-dataset-invalid.csv")
	invalidCSVContents := strings.Join([]string{
		"question,clip",
		`"This row should fail validation.","https://example.com/not-audio.txt"`,
		"",
	}, "\n")
	if err := os.WriteFile(invalidCSVPath, []byte(invalidCSVContents), 0o600); err != nil {
		fatal(fmt.Errorf("failed to write invalid CSV fixture: %w", err))
	}

	// Attempt an invalid upload first to confirm audio_url extension validation rejects non-audio URLs.
	if _, err := runAllowFailure(repoRoot, cliBinary, "aitaskbuilder", "dataset", "upload", "-d", datasetID, "-f", invalidCSVPath); err == nil {
		fatal(fmt.Errorf("expected invalid audio URL upload to fail"))
	} else {
		fmt.Println("Confirmed invalid audio URL upload was rejected.")
	}

	// Upload valid sample data containing supported audio URL extensions.
	if _, err := run(repoRoot, cliBinary, "aitaskbuilder", "dataset", "upload", "-d", datasetID, "-f", csvPath); err != nil {
		fatal(err)
	}

	// Check dataset status for extra visibility while running the manual flow.
	_, _ = runAllowFailure(repoRoot, cliBinary, "aitaskbuilder", "dataset", "check", "-d", datasetID)

	// Create a batch linked to the dataset and define the participant layout via batch_items.
	// batch_items replaces the deprecated standalone instructions endpoint, and this
	// manual flow intentionally references the audio_url dataset field even though
	// the checked-in OpenAPI file has not caught up with the backend yet.
	output, err = run(repoRoot, cliBinary,
		"aitaskbuilder", "batch", "create",
		"-n", "Audio URL Test Batch",
		"-w", workspaceID,
		"-d", datasetID,
		"--task-name", "Audio Review Task",
		"--task-introduction", "Listen to the audio clip and answer the question.",
		"--task-steps", "1. Listen to the clip\\n2. Answer the question",
		"--batch-items-json", batchItems,
	)
	if err != nil {
		fatal(err)
	}

	batchID, err := extractField(output, "ID")
	if err != nil {
		fatal(err)
	}
	fmt.Printf("Created batch: %s\n", batchID)

	// Set up the batch so the persisted preview route has a task group to open.
	if _, err := run(repoRoot, cliBinary,
		"aitaskbuilder", "batch", "setup",
		"-b", batchID,
		"-d", datasetID,
		"--tasks-per-group", "1",
	); err != nil {
		fatal(err)
	}

	// Preview the batch through the researcher preview URL flow.
	if _, err := run(repoRoot, cliBinary, "aitaskbuilder", "batch", "preview", batchID); err != nil {
		fatal(err)
	}

	fmt.Println("\nDone. Review output above for correctness.")
}

func repoRoot() (string, error) {
	_, filePath, _, ok := runtime.Caller(0)
	if !ok {
		return "", fmt.Errorf("failed to resolve script path")
	}

	return filepath.Abs(filepath.Join(filepath.Dir(filePath), "..", ".."))
}

func buildCLI(repoRoot, cliBinary string) error {
	fmt.Println("Building CLI...")
	cmd := exec.CommandContext(context.Background(), "go", "build", "-o", cliBinary, ".")
	_, err := runWithCheck(true, repoRoot, "go", cmd)
	return err
}

func run(repoRoot, cliBinary string, args ...string) (string, error) {
	// #nosec G702 -- cliBinary is the local binary built by this script, not shell-expanded user input.
	cmd := exec.CommandContext(context.Background(), cliBinary, args...)
	return runWithCheck(true, repoRoot, cliBinary, cmd)
}

func runAllowFailure(repoRoot, cliBinary string, args ...string) (string, error) {
	// #nosec G702 -- cliBinary is the local binary built by this script, not shell-expanded user input.
	cmd := exec.CommandContext(context.Background(), cliBinary, args...)
	return runWithCheck(false, repoRoot, cliBinary, cmd)
}

func runWithCheck(check bool, repoRoot, command string, cmd *exec.Cmd) (string, error) {
	fmt.Printf("\n$ %s %s\n", command, strings.Join(cmd.Args[1:], " "))

	cmd.Dir = repoRoot
	cmd.Env = os.Environ()

	output, err := cmd.CombinedOutput()
	text := string(output)
	fmt.Print(text)

	if err != nil && check {
		return text, fmt.Errorf("command failed: %w", err)
	}

	return text, err
}

func extractField(output, label string) (string, error) {
	matches := fieldPattern.FindAllStringSubmatch(output, -1)
	for _, match := range matches {
		if len(match) >= 3 && match[1] == label {
			return match[2], nil
		}
	}

	return "", fmt.Errorf("could not find %q in output", label)
}

func fatal(err error) {
	fmt.Fprintln(os.Stderr, err)
	os.Exit(1)
}
